### DISCUSSION TITLE
How to properly type useQuery's `meta`?

### DISCUSSION THREAD

**Discussion by denisborovikov** [TanStack/query#6045] [Q&A] [answered]
The `meta` type is inferred as `Record<string, unknown>`. Is there a way to type `useQuery` so the `meta` type is correctly inferred?

```js

useQuery({ queryFn: ({ meta }) =>  ... }) // meta = Record<string, unknown>
``` 

We can cast the type like this

```
interface Meta {
  id: string
}
...
const queryFn = ({ meta }) => {
  const { id } = meta as unknown as Meta
}
```

but is there something better than this?

**Comment by TkDodo** [comment, marked as answer]
we don't have a generic for `meta` so its type cannot be inferred. In v5, you can use module augmentation:

```
declare module '@tanstack/react-query' {
  interface Register {
    queryMeta: Meta
    mutationMeta: Record<string, unknown>
  }
}
```

**Reply by denisborovikov** [reply]
Oh good to know. Thank you as always!

**Reply by joel-daros** [reply]
Is there a way to do the same in V4?

**Reply by charliematters** [reply]
When I do this, all the types for react query imports break. I'm not well versed with module augmentation, but do you have any idea what I'm missing? The errors look like this: `Module '"@tanstack/react-query"' has no exported member 'QueryClient'.'
`

**Reply by bencergazda** [reply]
> When I do this, all the types for react query imports break.

The types should get fixed if you add an import or an export in the module declaration file.

```typescript
declare module '@tanstack/react-query' {
  interface Register {
    queryMeta: Meta
    mutationMeta: Record<string, unknown>
  }
}

export {};
```

**Reply by alexandre1921** [reply]
Is there a way to do the same in V5?

**Reply by TkDodo** [reply]
amending the `Register` interface _is_ the way in v5: https://tanstack.com/query/v5/docs/framework/react/typescript#registering-global-meta

**Reply by lkp-k** [reply]
is it possible to add meta when doing something like this? i'm using v5

```
queryClient.setQueryData(userQueryKey, result, { meta: { source: "remote" }})
```

**Reply by TkDodo** [reply]
nope

**Reply by lkp-k** [reply]
ah ok. ty.

is there a suggested way to note the source? in my case i have a RN app where we load from cache first, and if we're online we'll make the network call. Besides storing `{ source: DataSource; data: unknown }` or something to that effect, is there a more "react-query way" of doing this? thank you

**Reply by TkDodo** [reply]
make it part of your `data`

**Reply by JoaoBrlt** [reply]
Is it possible to use module augmentation to define the type for `meta` that depends on query/mutation inputs or output?
It would be nice if, for example, we want to implement `invalidates`, `errorMessage`, `successMessage`, etc. meta options that optionally are defined using a function that takes the mutations options as an argument.

**Comment by stevebeauge** [comment]
> we don't have a generic for meta

Is there any chance to add such generic (meaning directly in the lib, not relying on module augmentation)?

**Reply by TkDodo** [reply]
not planned, no
